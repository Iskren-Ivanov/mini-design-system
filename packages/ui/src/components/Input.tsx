import React from 'react';
import { cn } from '../utils/cn';

export interface InputProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
  variant?: 'default' | 'filled' | 'outline';
  size?: 'sm' | 'md' | 'lg';
  error?: boolean;
  label?: string;
  helperText?: string;
  errorText?: string;
}

const inputVariants = {
  default: 'border-[var(--color-border)] bg-[var(--color-bg-primary)] text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)]',
  filled: 'border-transparent bg-[var(--color-bg-tertiary)] text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)]',
  outline: 'border-2 border-[var(--color-border)] bg-transparent text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)]',
};

const inputSizes = {
  sm: 'px-3 py-1.5 text-sm',
  md: 'px-4 py-2 text-base',
  lg: 'px-4 py-3 text-lg',
};

export function Input({
  variant = 'default',
  size = 'md',
  error = false,
  label,
  helperText,
  errorText,
  className,
  id,
  ...props
}: InputProps) {
  const hasError = error || !!errorText;

  return (
    <div className='w-full'>
      {label && (
        <label
          htmlFor={id}
          className='block text-sm font-medium text-[var(--color-text-secondary)] mb-1'
        >
          {label}
        </label>
      )}
      <input
        id={id}
        className={cn(
          'w-full rounded-md border transition-colors',
          'placeholder:text-[var(--color-text-tertiary)]',
          'focus:outline-none focus:ring-2 focus:ring-offset-0',
          'disabled:opacity-50 disabled:cursor-not-allowed',
          hasError && 'border-[var(--color-error)] bg-[var(--color-error-bg)] focus:border-[var(--color-error)] focus:ring-[var(--color-error)]',
          !hasError && inputVariants[variant],
          inputSizes[size],
          className
        )}
        {...props}
      />
      {(helperText || errorText) && (
        <p className={cn(
          'mt-1 text-sm',
          hasError ? 'text-[var(--color-error)]' : 'text-[var(--color-text-tertiary)]'
        )}>
          {errorText || helperText}
        </p>
      )}
    </div>
  );
}
