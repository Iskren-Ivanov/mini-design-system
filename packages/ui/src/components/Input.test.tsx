import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { Input } from './Input';

describe('Input', () => {
  it('renders the input with a label and placeholder', () => {
    render(<Input label='Username' placeholder='Enter your username' id='username' />);
    expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/enter your username/i)).toBeInTheDocument();
  });

  it('renders without label when not provided', () => {
    render(<Input placeholder='Enter text' />);
    expect(screen.getByPlaceholderText(/enter text/i)).toBeInTheDocument();
  });

  it('displays helper text', () => {
    render(<Input helperText='This is a helper message' />);
    expect(screen.getByText(/this is a helper message/i)).toBeInTheDocument();
  });

  it('displays error text when error prop is true', () => {
    render(<Input error errorText='This field is required' />);
    expect(screen.getByText(/this field is required/i)).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toHaveClass('border-[var(--color-error)]');
  });

  it('displays error text when errorText prop is provided without error prop', () => {
    render(<Input errorText='Invalid input' />);
    expect(screen.getByText(/invalid input/i)).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toHaveClass('border-[var(--color-error)]');
  });

  it('updates value on change', () => {
    render(<Input label='Name' id='name' />);
    const input = screen.getByLabelText(/name/i) as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'John Doe' } });
    expect(input.value).toBe('John Doe');
  });

  it('is disabled when disabled prop is true', () => {
    render(<Input disabled placeholder='Disabled input' />);
    const input = screen.getByPlaceholderText(/disabled input/i);
    expect(input).toBeDisabled();
  });

  it('applies default variant styles', () => {
    render(<Input variant='default' />);
    expect(screen.getByRole('textbox')).toHaveClass('border-[var(--color-border)]');
  });

  it('applies filled variant styles', () => {
    render(<Input variant='filled' />);
    expect(screen.getByRole('textbox')).toHaveClass('bg-[var(--color-bg-tertiary)]');
  });

  it('applies outline variant styles', () => {
    render(<Input variant='outline' />);
    expect(screen.getByRole('textbox')).toHaveClass('border-2 border-[var(--color-border)]');
  });

  it('applies small size styles', () => {
    render(<Input size='sm' />);
    expect(screen.getByRole('textbox')).toHaveClass('px-3 py-1.5 text-sm');
  });

  it('applies large size styles', () => {
    render(<Input size='lg' />);
    expect(screen.getByRole('textbox')).toHaveClass('px-4 py-3 text-lg');
  });

  it('applies custom className', () => {
    render(<Input className='custom-class' />);
    expect(screen.getByRole('textbox')).toHaveClass('custom-class');
  });


  it('handles focus and blur events', () => {
    const handleFocus = vi.fn();
    const handleBlur = vi.fn();
    render(<Input onFocus={handleFocus} onBlur={handleBlur} />);
    const input = screen.getByRole('textbox');

    fireEvent.focus(input);
    expect(handleFocus).toHaveBeenCalledTimes(1);

    fireEvent.blur(input);
    expect(handleBlur).toHaveBeenCalledTimes(1);
  });

  // Snapshot tests
  it('matches snapshot for default variant', () => {
    const { asFragment } = render(<Input placeholder='Default input' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot for filled variant', () => {
    const { asFragment } = render(<Input variant='filled' placeholder='Filled input' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot for outline variant', () => {
    const { asFragment } = render(<Input variant='outline' placeholder='Outline input' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with label', () => {
    const { asFragment } = render(<Input label='Email' placeholder='Enter email' id='email' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with error state', () => {
    const { asFragment } = render(<Input error errorText='This field is required' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with helper text', () => {
    const { asFragment } = render(<Input helperText='This is a helper message' />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot for disabled state', () => {
    const { asFragment } = render(<Input disabled placeholder='Disabled input' />);
    expect(asFragment()).toMatchSnapshot();
  });
});
