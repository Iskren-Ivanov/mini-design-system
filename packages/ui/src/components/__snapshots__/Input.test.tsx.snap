// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > matches snapshot for default variant 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <input
      class="w-full rounded-md border transition-colors placeholder:text-[var(--color-text-tertiary)] focus:outline-none focus:ring-2 focus:ring-offset-0 disabled:opacity-50 disabled:cursor-not-allowed border-[var(--color-border)] bg-[var(--color-bg-primary)] text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)] px-4 py-2 text-base"
      placeholder="Default input"
    />
  </div>
</DocumentFragment>
`;

exports[`Input > matches snapshot for disabled state 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <input
      class="w-full rounded-md border transition-colors placeholder:text-[var(--color-text-tertiary)] focus:outline-none focus:ring-2 focus:ring-offset-0 disabled:opacity-50 disabled:cursor-not-allowed border-[var(--color-border)] bg-[var(--color-bg-primary)] text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)] px-4 py-2 text-base"
      disabled=""
      placeholder="Disabled input"
    />
  </div>
</DocumentFragment>
`;

exports[`Input > matches snapshot for filled variant 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <input
      class="w-full rounded-md border transition-colors placeholder:text-[var(--color-text-tertiary)] focus:outline-none focus:ring-2 focus:ring-offset-0 disabled:opacity-50 disabled:cursor-not-allowed border-transparent bg-[var(--color-bg-tertiary)] text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)] px-4 py-2 text-base"
      placeholder="Filled input"
    />
  </div>
</DocumentFragment>
`;

exports[`Input > matches snapshot for outline variant 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <input
      class="w-full rounded-md transition-colors placeholder:text-[var(--color-text-tertiary)] focus:outline-none focus:ring-2 focus:ring-offset-0 disabled:opacity-50 disabled:cursor-not-allowed border-2 border-[var(--color-border)] bg-transparent text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)] px-4 py-2 text-base"
      placeholder="Outline input"
    />
  </div>
</DocumentFragment>
`;

exports[`Input > matches snapshot with error state 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <input
      class="w-full rounded-md border transition-colors placeholder:text-[var(--color-text-tertiary)] focus:outline-none focus:ring-2 focus:ring-offset-0 disabled:opacity-50 disabled:cursor-not-allowed border-[var(--color-error)] bg-[var(--color-error-bg)] focus:border-[var(--color-error)] focus:ring-[var(--color-error)] px-4 py-2 text-base"
    />
    <p
      class="mt-1 text-sm text-[var(--color-error)]"
    >
      This field is required
    </p>
  </div>
</DocumentFragment>
`;

exports[`Input > matches snapshot with helper text 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <input
      class="w-full rounded-md border transition-colors placeholder:text-[var(--color-text-tertiary)] focus:outline-none focus:ring-2 focus:ring-offset-0 disabled:opacity-50 disabled:cursor-not-allowed border-[var(--color-border)] bg-[var(--color-bg-primary)] text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)] px-4 py-2 text-base"
    />
    <p
      class="mt-1 text-sm text-[var(--color-text-tertiary)]"
    >
      This is a helper message
    </p>
  </div>
</DocumentFragment>
`;

exports[`Input > matches snapshot with label 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium text-[var(--color-text-secondary)] mb-1"
      for="email"
    >
      Email
    </label>
    <input
      class="w-full rounded-md border transition-colors placeholder:text-[var(--color-text-tertiary)] focus:outline-none focus:ring-2 focus:ring-offset-0 disabled:opacity-50 disabled:cursor-not-allowed border-[var(--color-border)] bg-[var(--color-bg-primary)] text-[var(--color-text-primary)] focus:border-[var(--color-primary)] focus:ring-[var(--color-primary)] px-4 py-2 text-base"
      id="email"
      placeholder="Enter email"
    />
  </div>
</DocumentFragment>
`;
